src/main/java/spoon/support/compiler/jdt/ReferenceBuilder.java

```diff
@@ -99,6 +101,9 @@
 
 public class ReferenceBuilder {
 
+	// Allow to detect circular references and to avoid endless recursivity
+	// when resolving parameterizedTypes (e.g. Enum<E extends Enum<E>>)
+	private Set<TypeBinding> exploringParameterizedBindings = new HashSet<>();
```

Comment:
I have heard that spoon elements does not work well in HashSet/Map, because their implementation of #hashCode and #equals doesn't fit to HashSet contract. I do not know details. I am using IdentityHashMap or HashSet<String> on unique identificator of the elements. Please discuss the details with @monperrus if needed.